File: 2385. Amount of Time for Binary Tree to Be Infected.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap <Integer, Integer> hm = new HashMap<>();
    int ans = 0;
    public int amountOfTime(TreeNode root, int start) {
        dfs1(root);
        dfs2(root, start, 0);
        return ans;
    }
    int dfs1(TreeNode root){
        if(root == null) return 0;
        int val = Math.max(dfs1(root.left),dfs1(root.right));
        hm.put(root.val, val);
        return val + 1;
    }
    void dfs2(TreeNode  root, int target, int parVal) {
        if(root == null) return;
        if(root.val == target){
            ans = Math.max(parVal, hm.get(root.val));
            return;
        }
        if(root.left != null) {
            int right = 0;
            if(root.right != null) right = hm.get(root.right.val) + 2;
            dfs2(root.left, target, Math.max(parVal + 1, right));
        }
        if(root.right != null) {
            int left = 0;
            if(root.left != null) left = hm.get(root.left.val) + 2;
            dfs2(root.right, target, Math.max(parVal + 1, left));
        }
    }

    
}

File: 2385. Amount of Time for Binary Tree to Be Infected.cpp
/*
problem credits: https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/description/

You are given the root of a binary tree with unique values, and an integer start. At minute 0, an infection starts from the node with value start.

Each minute, a node becomes infected if:

The node is currently uninfected.
The node is adjacent to an infected node.
Return the number of minutes needed for the entire tree to be infected.

Solution:
    Do dfs and once start is in sub array, return true and max height from start to other part of the child
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int ans = 0;
public:
    pair<bool, int> dfs(TreeNode* root, int start) {
        if (root == nullptr) return {false, 0};
        pair<bool, int> left = dfs(root->left, start);
        pair<bool, int> right = dfs(root->right, start);

        if (root -> val == start) {
            ans = max(left.second, right.second);
            return {true, 0};
        }

        if (left.first) {
            ans = max(ans, left.second + right.second + 1);
            return {true, left.second+1};
        } else if (right.first) {
            ans = max(ans, right.second + left.second + 1);
            return {true, right.second+1};
        }
        return {false, max(left.second, right.second) + 1};
    }

    int amountOfTime(TreeNode* root, int start) {
        dfs(root, start);
        return ans;
    }
};